### PR TITLE
Feature/freeze phase 1 option2

### PR DESCRIPTION
**Pull request by denis-yuen**
To support the concept of immutable versions, two prototypes were created:
1) The first prototype took advantage of row level security in Postgres, essentially one can define conditions (like a frozen column) to denote that a row should not be changed. See https://github.com/dockstore/dockstore/compare/feature/freeze_phase_1?expand=1
2) This prototype uses JPA callback annotations to detect changes and reject them. 

There are advantages and disadvantages to each idea 
- Postgres row level security 
  - Pros
    - enforced at the database level, so mistakes at the JPA layer cannot override it
    - simple to work with once setup (DB will just ignore changes silently) 
      - the JPA approach will throw an exception, ending things hard when changes on a frozen version 
are attempted
      - this is incomplete, SQL statements issues against the DB will ignore changes silently. But JPA still notices something amiss and will fail hard
  - Cons 
    - we need to switch ownership of the DB (which should be done anyway)
    - but database migration might be tricky 
    - the simplest approach (frozen = no changes) means that once something is frozen, there really is no way to unfreeze a version, i.e. the webservice user can freeze something and actually have no way to unfreeze it whatsoever
    - sql policy statements can be complicated to understand (and have to be done totally outside of JPA)
      - along these times, I haven't locked down source files ... although there probably is a way to do it

- JPA callback annotations approach 
  - Pros
    - database is unchanged
    - the @PrePersist and @PostLoad annotations are easy to work with once setup
    - can code up unfreezing easily if needed
  - Cons 
    - the whole Hibernate/JPA lifecycle is potentially changed with [JPA Integrator](https://docs.jboss.org/hibernate/orm/5.2/javadocs/org/hibernate/jpa/event/spi/JpaIntegrator.html) to have different behaviour, we may have unusual database behaviour while we debug
    - for example, the default JPAIntegrator breaks a pile of our tests, so we need to use a slightly hacked version to maintain our understanding of JPA flushing for example
   - attempting to modify a frozen version leads to an exception (rather than being ignored)

### CODE REVIEW

**Comment by coverbeck** [comment]
Did we say frozen versions can be deleted? I think so, although the user might have to go through extra hoops to do so, i.e., it should be possible but not easy. 

Asking as I'm looking at your PSQL row security, and wondering at first glance if it will be possible.

**Comment by denis-yuen** [comment]
> Did we say frozen versions can be deleted? I think so, although the user might have to go through extra hoops to do so, i.e., it should be possible but not easy.
> 
> Asking as I'm looking at your PSQL row security, and wondering at first glance if it will be possible.

Looking at David's "Immutable Versions" doc, I see

> We use “immutable” somewhat loosely here, to mean that we propose a specific kind of artifact from workflows that should not be alterable (other than perhaps removal). 

https://github.com/dockstore/dockstore/issues/1995#issuecomment-502836369 shows delete buttons

So the intent might have been, yes, delete should be possible. But not easy.  

This doesn't rule out using row security, it might be possible to alter the row security condition to be a bit less naive at the cost of more complexity.

**Comment by coverbeck** [comment]
It feels like the DB approach is the right one. In addition to the case you cited of potential JPA errors, there is also direct database access (migrations, somebody running psql), and (maybe?) the possibility of there someday being additional services running outside of the webservice (I keep hearing Gary say microservices :) ).

I don't know SQL policy statements (one of your cons). I googled around, and it seems another way you could do it is with [triggers](https://dba.stackexchange.com/questions/74517/how-to-create-a-postgres-trigger-to-ensure-field-is-non-updateable). Any reason you went with policies vs. triggers? I don't know enough to have an opinion on which approach is better.

If we do go the DB route, seems like we should figure out how to revert/delete for extreme cases. Presumably policies (and triggers) can be overridden on a per user basis?

File: dockstore-webservice/src/main/java/io/dockstore/webservice/CustomJPAIntegrator.java [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+     * @param sessionFactory  The session factory being created
+     * @param serviceRegistry The session factory's service registry
+     */
+    public void integrate(Metadata metadata, SessionFactoryImplementor sessionFactory, SessionFactoryServiceRegistry serviceRegistry) {
+
+        // first, register the JPA-specific persist cascade style
+        try {
+            oldPersistCascadeStyle = CascadeStyles.getCascadeStyle("persist");
+        } catch (Exception e) {
+            assert true;
```
[the remark this answers is not in the document]
Comment (denis-yuen) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/CustomJPAIntegrator.java]: this class is mostly copied from org.hibernate.jpa.event.spi.JpaIntegrator, don't actually know what this part is

**Pull request closed without merging**